onnx/defs/sequence/defs.cc

```diff
@@ -634,4 +634,297 @@ ONNX_OPERATOR_SET_SCHEMA(
+be a sequence. The length of the first input sequence will determine the number of samples in the
+outputs. The number of inputs and outputs, should match the one of the subgraph.
+
+For each element in the output, a sample will be extracted from the input sequence(s) and the sub-graph
+will be applied to it. The outputs will contain the outputs of the sub-graph for each sample.)DOC";
+
+void SequenceMapInferenceFunction(InferenceContext& ctx) {
+  auto num_inputs = ctx.getNumInputs();
+  assert(num_inputs > 0);
+
+  auto num_outputs = ctx.getNumOutputs();
+  assert(num_outputs > 0);
+
+  std::vector<TypeProto> tmp_type_protos;
+  std::vector<const TypeProto*> subgraph_input_types;
+  subgraph_input_types.reserve(num_inputs);
+  for (size_t inputIndex = 0; inputIndex < num_inputs; inputIndex++) {
+    auto input_type = ctx.getInputType(inputIndex);
+    if (input_type == nullptr) {
+      fail_type_inference("Input ", inputIndex, " expected to have type info");
+    }
+    if (input_type->value_case() == TypeProto::kSequenceType) {
+      tmp_type_protos.emplace_back();
+      tmp_type_protos.back().CopyFrom(input_type->sequence_type().elem_type());
+      subgraph_input_types.push_back(&tmp_type_protos.back());
+    } else {
+      if (inputIndex == 0)
+        fail_type_inference("Input ", inputIndex, " expected to be a sequence type");
+      subgraph_input_types.push_back(input_type);
+    }
+  }
+
+  GraphInferencer* graphInferencer = ctx.getGraphAttributeInferencer("body");
+  if (graphInferencer) {
+    std::vector<const TensorProto*> input_data(num_inputs, nullptr);
+    std::vector<const TypeProto*> subgraph_output_types =
+      graphInferencer->doInferencing(subgraph_input_types, input_data);
+
+    // if empty(), assume inferencing was skipped
+    if (!subgraph_output_types.empty()) {
+      if (subgraph_output_types.size() != num_outputs) {
+        fail_type_inference(
+            "Graph attribute inferencing returned type information for ",
+            subgraph_output_types.size(), " outputs. Expected ", num_outputs);
+      }
+
+      for (size_t outputIndex = 0; outputIndex < num_outputs; outputIndex++) {
+        auto* subgraph_output_type = subgraph_output_types[outputIndex];
+        ctx.getOutputType(outputIndex)
+            ->mutable_sequence_type()
+            ->mutable_elem_type()
+            ->CopyFrom(*subgraph_output_type);
+      }
+    }
+  }
+}
+
+bool BuildSequenceMapBodyFunc(const FunctionBodyBuildContext& ctx,
+                              const OpSchema& schema,
+                              FunctionProto& functionProto) {
+  auto body_attr = ctx.getAttribute("body");
+  if (!body_attr || !body_attr->has_g())
+    throw std::invalid_argument("Invalid ``body`` argument. Expected a graph");
+  const GraphProto& body = body_attr->g();
+
+  auto g_inputs = body.input();
+  auto g_outputs = body.output();
+
+  auto node_inputs = ctx.getInputs();
```

Comment:
I suspect that instead of using the actual input-parameter names, it would be necessary to introduce (formal) inputs to the FunctionProto, and use those names. These names can be created (eg., "input1", "input2", etc.) as long as we avoid name collisions.

Reply:
Good point. Done